docs/examples/langchain/langchain_chat_pandas_df.py

```diff
@@ -0,0 +1,182 @@
+
+    show_chain_of_thought = param.Boolean(default=False)
+
+    def _get_agent_message(self, message: str) -> pn.chat.ChatMessage:
+        return pn.chat.ChatMessage(message, user=self.user, avatar=self.avatar)
+
+
+class AppState(param.Parameterized):
+    data = param.DataFrame()
+
+    llm = param.Parameter(constant=True)
+    pandas_df_agent = param.Parameter(constant=True)
+
+    config: AgentConfig = param.ClassSelector(class_=AgentConfig)
+    environ: Environment = param.ClassSelector(class_=Environment, constant=True)
+
+    def __init__(
+        self, config: AgentConfig | None = None, environ: Environment | None = None
+    ):
+        if not config:
+            config = AgentConfig()
+        if not environ:
+            environ = Environment()
+        super().__init__(config=config, environ=environ)
+
+    @param.depends("environ.OPENAI_API_KEY", on_init=True, watch=True)
+    def _reset_llm(self):
+        with param.edit_constant(self):
+            if self.environ.OPENAI_API_KEY:
+                self.llm = ChatOpenAI(
+                    temperature=0,
+                    model="gpt-3.5-turbo-0613",
+                    api_key=self.environ.OPENAI_API_KEY,
+                    streaming=True,
+                )
+            else:
+                self.llm = None
+
+    @param.depends("llm", "data", on_init=True, watch=True)
+    def _reset_pandas_df_agent(self):
+        with param.edit_constant(self):
+            if not self.error_message:
+                self.pandas_df_agent = create_pandas_dataframe_agent(
+                    self.llm,
+                    self.data,
+                    verbose=True,
+                    agent_type=AgentType.OPENAI_FUNCTIONS,
+                    handle_parsing_errors=True,
+                )
+            else:
+                self.pandas_df_agent = None
+
+    @property
+    def error_message(self):
+        if not self.llm and self.data is None:
+            return dedent(
+                """\
+                Please provide your `OPENAI_API_KEY`, **upload a `.csv` file** 
+                and click the **send** button."""
+            )
+        if not self.llm:
+            return "Please provide your `OPENAI_API_KEY`."
+        if self.data is None:
+            return "Please **upload a `.csv` file** and click the **send** button."
+        return ""
+
+    @property
+    def welcome_message(self):
+        text = dedent(
+            f"""
+            I'm your <a href="\
+            https://python.langchain.com/docs/integrations/toolkits/pandas" target="_blank">\
+            LangChain Pandas DataFrame Agent</a>.
+                    
+            I execute LLM generated Python code under the hood - this can be bad if the `llm`
+            generated Python code is harmful. Use cautiously!
+
+            {self.error_message}"""
+        ).strip()
+        if self.data is None:
+            text += f"""
+
+Example: <a href="{PENGUINS_URL}" download>penguins.csv<a>
+"""
```

Comment:
Can dedent also be used here?

Reply:
Done. I do it for you 😄 . I'm not a big fan of dedent as 1) it adds a tiny performance cost 2) it adds two extra code lines 3) it makes the code extra indented causing issues with max characters in line.

Reply:
Thanks.

I think the tiny performance cost in the grand scheme of things is negligible; if you'd like to avoid the tiny performance cost, you can use parentheses + individual quotes, e.g. instead of...
```
some_multiline_str = dedent(
    """
    line 1
    line 2
    """
)
```


```
some_multiline_str = (
    "line 1\n"
    "line 2\n"
)
```

Anyhow, I personally believe code readability is more important; without the dedent (or the aforementioned method), it's distracting to see code unindented within functions.